packages/codecs-core/src/offset-codec.ts

```diff
@@ -0,0 +1,83 @@
+import { assertByteArrayOffsetIsNotOutOfRange } from './assertions';
+import { Codec, createDecoder, createEncoder, Decoder, Encoder, Offset } from './codec';
+import { combineCodec } from './combine-codec';
+
+type OffsetConfig = {
+    postOffset?: PostOffsetFunction;
+    preOffset?: PreOffsetFunction;
+};
+
+type PreOffsetFunctionScope = {
+    /** The entire byte array. */
+    bytes: Uint8Array;
+    /** The original offset prior to encode or decode. */
+    preOffset: Offset;
+    /** Wraps the offset to the byte array length. */
+    wrapBytes: (offset: Offset) => Offset;
+};
+
+type PreOffsetFunction = (scope: PreOffsetFunctionScope) => Offset;
+type PostOffsetFunction = (
+    scope: PreOffsetFunctionScope & {
+        /** The modified offset used to encode or decode. */
+        newPreOffset: Offset;
+        /** The original offset returned by the encoder or decoder. */
+        postOffset: Offset;
+    },
+) => Offset;
+
+/**
+ * Moves the offset of a given encoder.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function offsetEncoder<TEncoder extends Encoder<any>>(encoder: TEncoder, config: OffsetConfig): TEncoder {
+    return createEncoder({
+        ...encoder,
+        write: (value, bytes, preOffset) => {
+            const wrapBytes = (offset: Offset) => modulo(offset, bytes.length);
+            const newPreOffset = config.preOffset ? config.preOffset({ bytes, preOffset, wrapBytes }) : preOffset;
+            assertByteArrayOffsetIsNotOutOfRange('offsetEncoder', newPreOffset, bytes.length);
+            const postOffset = encoder.write(value, bytes, newPreOffset);
+            const newPostOffset = config.postOffset
+                ? config.postOffset({ bytes, newPreOffset, postOffset, preOffset, wrapBytes })
+                : postOffset;
+            assertByteArrayOffsetIsNotOutOfRange('offsetEncoder', newPostOffset, bytes.length);
+            return newPostOffset;
+        },
```

Comment:
Small up-to-you nit: I think it's a little more clear if you plumb `offset` through this block, as defined by the `createEncoder` API. Same thing below with `offsetDecoder`. It makes it a bit clearer to see that you're using both `preOffset` and `postOffset` to slide the actual `offset` along.

Reply:
Hmm tbh, as we're always introducing 4 new terms to describe offsets (`preOffset`, `newPreOffset`, `postOffset` and `newPostOffset`), I'd rather be extra-explicit here and avoid having a generic `offset` attribute.

Reply:
A generic attribute isn't what I meant. But nbd, merged.